call_option_test.go

```diff
@@ -1,47 +1,86 @@
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//     * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
 package gax
 
 import (
-	"reflect"
 	"testing"
 	"time"
 
+	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 )
 
-func TestCallOptions(t *testing.T) {
-	expected := &CallSettings{
-		time.Second * 1,
-		RetrySettings{
-			map[codes.Code]bool{codes.Unavailable: true, codes.DeadlineExceeded: true},
-			BackoffSettings{
-				MultipliableDuration{time.Second * 2, time.Second * 4, 3.0},
-				MultipliableDuration{time.Second * 5, time.Second * 7, 6.0},
-			},
-		},
-	}
+func TestBackofDefault(t *testing.T) {
+	backoff := Backoff{}
 
-	settings := &CallSettings{}
-	opts := []CallOption{
-		WithTimeout(time.Second * 1),
-		WithRetryCodes([]codes.Code{codes.Unavailable, codes.DeadlineExceeded}),
-		WithDelayTimeoutSettings(time.Second*2, time.Second*4, 3.0),
-		WithRPCTimeoutSettings(time.Second*5, time.Second*7, 6.0),
+	want := []time.Duration{1, 2, 4, 8, 16, 30, 30, 30, 30, 30}
+	for i, w := range want {
+		want[i] = w * time.Second
 	}
-	callOptions(opts).Resolve(settings)
 
-	if !reflect.DeepEqual(settings, expected) {
-		t.Errorf("piece-by-piece settings don't match their expected configuration")
+	for _, w := range want {
+		if d := backoff.pause(); d > w {
+			t.Errorf("Backoff duration should be at most %s, got %s", w, d)
+		}
 	}
+}
 
-	settings = &CallSettings{}
-	expected.Resolve(settings)
+func TestBackoffExponential(t *testing.T) {
+	backoff := Backoff{Initial: 1, Max: 20, Mult: 2}
```

Comment:
My fault. I suggested using a struct instead of a New function, since Max and Mult would likely be defaulted.

If you think a function is better, fine. It will be seldom used in any case.